### PR TITLE
Implement read-only, write-only and required properties

### LINKED ISSUE #226: Implementing readOnly, writeOnly and Required Properties
### I'm submitting a 
- [ ] bug report.
- [X] feature request.

### Current Behaviour:

Right now although the doc contains properties like readOnly ,writeOnly , Required the functionality is not implemented in Hydrus. We need to implement this functionality so that we can have checksums instead of pk's or UUID's , advantage of that would be that client can reconstruct the checksum to check what is to be fetched .
### Do you want to work on this issue?

~YES~

EDITED BY Mec-iS

### PR DESCRIPTION
**Pull request by vddesai1871**
Fixes #226 

### Checklist
- [x] My branch is up-to-date with upstream/develop branch.
- [x] Everything works and tested for Python 3.5.2 and above.

### Description

Implemented read-only, write-only and required properties.
### Change logs

Added code to handle read-only, write-only and required properties.
Added tests for write-only and required-props. 

Adding tests for read-only props requires major changes in test file, which involves #317 and other changes to generations of dummy objects respecting read-only constraint and create appropriate objects of PUT (should have read-only props) and POST(should not have read-only props), and adapting to those changes in almost every individual test. So I will add that in another PR after discussing. (We have #317 open so we won't miss out adding this test).

### CODE REVIEW

**Comment by chrizandr** [comment]
Will check this out tomorrow in detail.

**Comment by vddesai1871** [comment]
>A better approach would be to store read-only, write-only and required values along with the property definitions which are created when the initial db schema is generated and all the classes and properties are added to the database.

I had to pass api_doc to crud.py for implementing write-only props. Read-only and required works without any modification in crud.py. So I think we will have to just store write-only pros in db.

But we are also using doc for segragation of objects in responses. 
(ie. 
```
            if doc.collections[collection]["collection"].class_.path == inst_class_name:
                nested_class_path = doc.collections[collection]["collection"].path
```
)
How can we achive this without passing api_doc?

**Comment by vddesai1871** [comment]
>I don't remember how things are implemented is read-only, write-only stored in the database. How can these work but not the required. Can you explain a bit

Read-only and required work without crud.py but write-only don't.
I am handling this constraint as they occur, so If a client tries to create a new object or update an old object ( he must set the value of all required properties in request data(object)), so I am checking that in resources.py (in `put` and `post` methods).
Same goes for read-only properties handled in resources.py.
For write-only properties, the server is supposed to do not add write-only properties and their values in response, and response object is being created in crud.py so I've written code in crud.py to skip fetching and adding write-only properties in the response object.(We can also do this in resources.py by deleting write-only properties from the response object created by crud.py before sending it to the client, but I thought that won't be efficient)

**Comment by xadahiya** [comment]
@Mec-iS  will review in detail and merge over the weekend. This is a crucial part.

**Comment by xadahiya** [comment]
@vddesai1871 Explain me this, why aren't we checking for all 3 `read-only`, `write-only` and `required` types in `crud.py` itself? Whenever tests fail we can just raise an Error and won't need to change anything else. Also, this will make the `param` of all functions in crud equivalent (right now some have doc as parameter and some don't). 
I can't see the benefit of checking `write-only` in crud.py and the rest in resources.py. Things should be consistent don't you think?

**Comment by vddesai1871** [comment]
>I can't see the benefit of checking write-only in crud.py and the rest in resources.py. Things should be consistent don't you think?

@xadahiya I believe handling write-only and other two as the same would be an over-generalization, by a superficial look they do seem closely related but by their essence or by their purpose, they are very different things. Write-only has nothing to do with raising Error or to do anything with request body, it just provides a requirement regarding how the response body will be created( suggesting which properties to exclude from response body). While read-only and required provides information about what is considered as a valid request body and what's not.  And their might be a violation of this desired request body schema and we must handle that by raising some sort of error or exception. **In brief write-only provides constraint on response body and other two provide constriants on request body and that's why they are being handled differently.**

If we were creating a normal 'primitive' API we wouldn't have other choice but to add these fields in db and thus check for them in CRUD, going a step forward we could have used some sort of request serializer like marshmallow to define request-response schema and handle these constriaints at view level. But with hydra we already have api doc (the schema) and we just have to write validation code.

PROPOSAL: Leave crud as it is and handle everything at `view` level in resources.py.
1) Read-only and required will remain same,
2) Create a function to finalize the response data, which will remove write-only data from object returned by crud that's how it (dump-only misnomer for write-only) is being handled in marshmallow and other serializers, the same function would also handle path of nested objects. ( crud will set {`property_name`: id}, the function with help of api_doc will convert it to {`property_name`: actual_path[/id]})
We won't need to pass api_doc to crud, 
Change in api_doc won't require changes in db(If a property's write-only or anyother attribute is changed hydrus will work fine without having to change anything in db).

**Comment by chrizandr** [comment]
@vddesai1871 I agree with your argument, but @xadahiya is right. Both request and responses are handled in crud at the lower levels. The hierarchy is `View -> Crud -> DB`, for both request and response objects. Like you said, we can implement the constraints in `View`, but it can also be implemented in `Crud`. I think we need to choose one only, not have a mix of both. It would be better to have it either in `Crud` or in `View`, but not separate for request and responses.

**Comment by vddesai1871** [comment]
> I think we need to choose one only, not have a mix of both. It would be better to have it either in Crud or in View, but not separate for request and responses.

@chrizandr I agree, that's why I proposed handling write-only also in `View` in my latest proposal.

**Comment by vddesai1871** [comment]
>I have one more idea that I had in mind for a while now, we could modify the API Doc object to do all these checks. Right now, there are a lot of nested ifs in the Views because of a large number of API Doc checks that are happening on the object. What if we have just a simple API_Doc.validate() method that checks all constraints and conditions that are required for an object to conform to the doc. This would remove a lot of the complicated code from views. 

I too had something similar in mind as part of `improvement and refactoring in hydrus` GSoC project :)

**Comment by xadahiya** [comment]
@vddesai1871 the latest proposal sounds good. I think you should proceed with this then we can merge it. 
Regarding the rest of the improvements I think we can do those during GSoC.

**Comment by Mec-iS** [comment]
@xadahiya @chrizandr 
Is this ok to merge?

File: hydrus/helpers.py [hunk 1]
```diff
@@ -103,3 +103,42 @@ def checkClassOp(class_type: str, method: str) -> bool:
         if supportedOp.method == method:
             return True
     return False
+
+
+def check_required_props(class_type: str, obj: Dict[str, Any]) -> bool:
+    for prop in get_doc().parsed_classes[class_type]["class"].supportedProperty:
```
Comment (Mec-iS) [inline hydrus/helpers.py]: docstring for the function is missing

File: hydrus/helpers.py [hunk 2]
```diff
@@ -103,3 +103,42 @@ def checkClassOp(class_type: str, method: str) -> bool:
...
+        if prop.required:
+            if prop.title not in obj:
+                return False
+    return True
+
+
+def check_read_only_props(class_type: str, obj: Dict[str, Any]) -> bool:
+    for prop in get_doc().parsed_classes[class_type]["class"].supportedProperty:
```
Comment (Mec-iS) [inline hydrus/helpers.py]: docstring for the function is missing

File: hydrus/helpers.py [hunk 3]
```diff
@@ -103,3 +103,42 @@ def checkClassOp(class_type: str, method: str) -> bool:
...
+        if prop.read:
+            if prop.title in obj:
+                return False
+    return True
+
+
+def get_nested_class_path(class_type: str) -> Tuple[str, bool]:
+    for collection in get_doc().collections:
```
Comment (Mec-iS) [inline hydrus/helpers.py]: docstrings missing

File: hydrus/helpers.py [hunk 4]
```diff
@@ -103,3 +103,42 @@ def checkClassOp(class_type: str, method: str) -> bool:
...
+        if get_doc().collections[collection]["collection"].class_.title == class_type:
+            return get_doc().collections[collection]["collection"].path, True
+
+    return get_doc().parsed_classes[class_type]["class"].path, False
+
+
+def finalize_response(class_type: str, obj: Dict[str, Any]) -> Dict[str, Any]:
+    for prop in get_doc().parsed_classes[class_type]["class"].supportedProperty:
```
Comment (Mec-iS) [inline hydrus/helpers.py]: docstring missing

**Comment by xadahiya** [comment]
We should merge this. @vddesai1871 Can you please resolve the conflicts?

**Pull request merged**